### ISSUE TITLE
Are we going to implement memswap_limit?

### ISSUE THREAD

**Issue by xinyifly** [aws/amazon-ecs-cli#307]
The `MemorySwap` is forced to twice of `Memory` at the moment. I may need `memswap_limit` to disable swap.

Reference:
https://docs.docker.com/compose/compose-file/compose-file-v2/#cpu_count-cpu_percent-cpu_shares-cpu_quota-cpus-cpuset-domainname-hostname-ipc-mac_address-mem_limit-memswap_limit-mem_swappiness-mem_reservation-oom_score_adj-privileged-read_only-shm_size-stdin_open-tty-user-working_dir

**Comment by SoManyHs** [comment]
Hi @xinyifly,

Currently, the ecs agent does not support `memory-swap` as a parameter (See; https://github.com/aws/amazon-ecs-agent/issues/771), which leads to the behavior you mentioned of `MemorySwap` being set to twice the limit specified in `Memory`. However, on the ECS Optimized AMI, the swappiness is not enabled by default, so while there is a limit being specified for `MemorySwap`, it's possible that it's not actually being used.

Until ECS decides to support memory-swap as part of our Task Definition, we are relying on the default behavior/settings of docker and your EC2 instance.

**Comment by pbecotte** [comment]
Hi @SoManyHs We had containers failing recently- it looks very much like it was trying to swap to the EBS volume even with that swappiness setting at 0 and swap disabled on the host. Do you know if this would be the case? If it is- is there a workaround, or a way to get this ticket re-opened?

**Comment by SoManyHs** [comment]
Hi @pbecotte,

Are you setting this under the memswap_limit field or mem_swappiness? If you could provide your docker-compose file, that'd also be helpful -- depending on the field, it might be worth opening as a separate issue.

**Comment by pbecotte** [comment]
It is just an ECS config using the json format. We aren't setting it at all- I am just looking at what are (presumably) the defaults the ecs-agent is using. This is a container that has a 4g hard limit set in the ecs agent and no other memory setting specified-

```
$ docker inspect 123 | grep memory
            "Memory": 4194304000,
            "KernelMemory": 0,
            "MemoryReservation": 0,
            "MemorySwap": 8388608000,
            "MemorySwappiness": 0,
```

**Comment by SoManyHs** [comment]
Hm, if swap is disabled on the host, there shouldn't be any swapping happening. Is there any output you could provide that suggests that swapping with EBS was happening?

**Comment by xinyifly** [comment]
@SoManyHs This is a example `docker stats` output:
```
CONTAINER           CPU %               MEM USAGE / LIMIT     MEM %               NET I/O             BLOCK I/O           PIDS
52dd                0.06%               127.9 MiB / 128 MiB   99.92%              3.96 GB / 3.45 GB   785 GB / 164 kB     0
```

Notice the huge BLOCK I/O. In worse case, it can be several TB's in days. The effect is the BLOCK I/O will cause high load average.
```
 02:08:47 up 12 days, 19:14,  1 user,  load average: 0.24, 0.50, 0.62
USER     TTY      FROM              LOGIN@   IDLE   JCPU   PCPU WHAT
ec2-user pts/0    ip-xxx-xx-xx-xxx  01:40    0.00s  0.01s  0.00s w
```
(In this case, it's in low load average. But if it's not setting correctly. BLOCK I/O raises rapidly and over 20+ load average with low CPU usage)

**Comment by nmeyerhans** [comment]
@xinyifly The block IO you're seeing isn't caused by swapping, since you don't have swap enabled on the host. If swapping was occurring, you'd see a much larger amount of data being written to disk (the second field under the BLOCK I/O header), but you've only written 164 kB. The IO is caused by paging because your container's memory limit is too small to support a useful page cache and all file access ends up needing to read pages from the disk.

Basically this is an indication that your memory limit is too low to support this container's activity. There is no adjustment to swappiness that could reduce the block IO. You can address this in one of two ways:
1. Reduce the amount of memory that your application allocations, thus freeing up memory for a page cache.
2. Increase the memory limit for this container.

**Comment by SoManyHs** [comment]
Hi @xinyifly,

Thanks for the info! As @nmeyerhans mentioned, it does appear to be an issue with memory limit and not with swap settings. You can try tweaking the `mem_limit` field in your docker-compose.yml file for the container in question and see if that helps.

Let me know if you have further questions!

**Comment by pbecotte** [comment]
Well thank you so much @nmeyerhans your post finally led me down the right path to figure out what is causing our problems! I guess I never realized how effective the disk cache actually is at preventing reads from the filesystem lol.

So @SoManyHs - looks like I was just misinterpreting what I was seeing, thank you both for helping me figure this out.
